includes/Core/Consent_Mode/Consent_Mode.php

```diff
@@ -80,18 +80,64 @@ public function register() {
 	 * @since n.e.x.t
 	 */
 	protected function render_gtag_consent_snippet() {
-			$consent_defaults = array(
-				'ad_personalization' => 'denied',
-				'ad_storage'         => 'denied',
-				'ad_user_data'       => 'denied',
-				'analytics_storage'  => 'denied',
-				'regions'            => $this->consent_mode_settings->get_regions(),
-			);
-			?>
+		$consent_defaults = array(
+			'ad_personalization' => 'denied',
+			'ad_storage'         => 'denied',
+			'ad_user_data'       => 'denied',
+			'analytics_storage'  => 'denied',
+			'regions'            => $this->consent_mode_settings->get_regions(),
+			'wait_for_update'    => 500, // Allow 500ms for Consent Management Platforms (CMPs) to update the consent status.
+		);
+
+		$consent_category_map = apply_filters(
+			'googlesitekit_consent_category_map',
+			array(
+				'statistics' => array( 'analytics_storage' ),
+				'marketing'  => array( 'ad_storage', 'ad_user_data', 'ad_personalization' ),
+			)
+		);
+		// TODO: We should extract this JS and make it a bit more readable.
+		?>
 <!-- <?php echo esc_html__( 'Google tag (gtag.js) Consent Mode snippet added by Site Kit', 'google-site-kit' ); ?> -->
 <script id='google_gtagjs-js-consent-mode'>
 window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}
-gtag("consent","default", <?php echo wp_json_encode( $consent_defaults ); ?>);
+gtag("consent","default",<?php echo wp_json_encode( $consent_defaults ); ?>);
+window._googlesitekitConsentCategoryMap = <?php	echo wp_json_encode( $consent_category_map ); ?>;
+document.addEventListener('wp_listen_for_consent_change', function(event) {
+	if (event.detail) {
+		var consentParameters = Object.entries(event.detail).reduce((parameters,[category,status]) => {
+			if (window._googlesitekitConsentCategoryMap[category]) {
+				var mappedStatus = status === 'allow' ? 'granted' : 'denied';
+				window._googlesitekitConsentCategoryMap[category].forEach((parameter) => {
```

Comment:
Regarding your comment about `const` above, I assume you're referring to the language feature rather than saving a character, correct? JS features are something we should be mindful of (and something we could revisit in a follow-up issue all at once if we prefer), but things like arrow functions, `Object` methods and other methods should be evaluated for safety and ideally reimplemented to use safer alternatives. It may sound a bit silly, but there is no minimum standard for features used on the front end (WP core even has theme support which controls html5 output).

Reply:
Thanks, yes that's a good point, and something that occurred to me during development - I should have paid it more heed. I've rewritten the snippet to avoid any ES5+ language features. In time we should see about moving it out to a transpiled asset so we can treat it as any other code.